src/best-practices/caddy.rst

```diff
@@ -0,0 +1,167 @@
+The team's experience with using it in production has shown it to be superior
+for configuration and monitoring capabilities, as well as overall performance.
+
+CouchDB's sample haproxy configuration is present in the `code repository`_ and
+release tarball as ``rel/haproxy.cfg``.
+
+However, ``Caddy`` is a suitable alternative. Below are instructions on
+configuring Caddy appropriately.
+
+.. _HAProxy: http://haproxy.org/
+.. _code repository: https://github.com/apache/couchdb/blob/master/rel/haproxy.cfg
+
+Basic configuration
+===================
+
+Here's a basic excerpt from a Caddyfile in
+``/<path>/<to>/<site>/Caddyfile``. This will proxy all
+requests from ``http(s)://domain.com/...`` to ``http://localhost:5984/...``
+
+.. code-block:: text
+
+    domain.com {
+
+        import /path/to/other/config.caddy # logging, error handling etc.
+
+        proxy / localhost:5984 {
+            transparent
+        }
+
+    }
+
+Note that, because Caddy is https-by-default, you must explicitly include the
+``http://`` protocol in the site address if you do NOT want Caddy
+to automatically acquire and install an SSL certificate and begin accepting
+``https`` connections on port 443.
+
+Reverse proxying CouchDB in a subdirectory with Caddy
+=====================================================
+
+It can be useful to provide CouchDB as a subdirectory of your overall domain,
+especially to avoid CORS concerns. Here's an excerpt of a basic Caddy
+configuration that proxies the URL ``http(s)://domain.com/couchdb`` to
+``http://localhost:5984`` so that requests appended to the subdirectory, such
+as ``http(s)://domain.com/couchdb/db1/doc1`` are proxied to
+``http://localhost:5984/db1/doc1``.
+
+.. code-block:: text
+
+    domain.com {
+
+        import /path/to/other/config.caddy # logging, error handling etc.
+
+        proxy /couchdb localhost:5984 {
+            transparent
+            without /couchdb
+        }
+
+    }
+
+Note that in the above configuration, the *Verify Installation* link in
+Fauxton may not succeed.
+
+Reverse proxying + load balancing for CouchDB clusters
+======================================================
+
+Here's a basic excerpt from a Caddyfile in
+``/<path>/<to>/<site>/Caddyfile``. This will proxy and evenly distribute all
+requests from ``http(s)://domain.com/...`` among 3 CouchDB cluster nodes
+at ``localhost:15984``, ``localhost:25984`` and ``localhost:35984``.
+
+Caddy will check the status, i.e. health, of each node every 5 seconds;
+if a node goes down, Caddy will avoid proxying requests to that node until it
+comes back online.
+
+.. code-block:: text
+
+    domain.com {
+
+        import /path/to/other/config.caddy # logging, error handling etc.
+
+        proxy / http://localhost:15984 http://localhost:25984 http://localhost:35984 {
+            policy round_robin
+            health_check /_up
+            health_check_duration 5s
+            try_interval 500ms
+            timeout 1h
```

Comment:
Does this mean that after seeing a successful `200`, Caddy won't check again for another hour? That could be catastrophic in a cluster setting where nodes go offline.

Reply:
Turns out that this is actually not a good option performance-wise for CouchDB right now - and yes, we know this isn't ideal.

This is why we do [this](https://github.com/apache/couchdb/blob/master/rel/haproxy.cfg#L27) in the haproxy config.

Would you be willing to test changing this on one of your clusters?

Reply:
Confession: I've always had Caddy handle timeouts... 🙈 I use the default setting of 30s

The PR-example in question was really my attempt to comply with the existing haproxy config 'spec'; obviously I mistranslated* the `timeout server 3600000` in haproxy. 

So the haproxy config's `timeout client 150000`  i.e. 2.5 minutes setting is the recommended timeout duration?

> Turns out that this is actually not a good option performance-wise for CouchDB right now

\* In my defense, yeah, this ^ is kind weird :)

------------

Below is the _actual_ config I'm running in prod commented with definitions and some defaults included for added clarity.

```
proxy /db 127.0.0.1:5984 {$REMOTE_COUCH_ONE} {$REMOTE_COUCH_TWO} {

  # My custom LB settings [Caddy default setting in brackets]
  policy least_conn # Select backend with fewest active connections [random]
  fail_timeout 30s  # How long to remember a failed request to a backend [0s]
  max_fails 2       # Failed reqs within fail_timeout to consider a backend down [1]
  try_duration 10s  # How long client may hang as proxy looks for a backend [0s]

  # Caddy default LB settings (included for reference)
  timeout 30s         # duration before timing out the connection to upstream
  try_interval 250ms  # how long to wait between selecting another upstream host

  # Health check settings
  health_check /_up         # ...
  health_check_interval 10s # ...
  health_check_timeout 15s  # Deadline to respond, otherwise req is a failure

  # Header settings for requests to CouchDB
  transparent
  header_upstream X-Forwarded-Ssl on
  header_upstream -Authorization

  # Header settings for response back to client
  header_downstream X-Lol-Its {$API_INSTANCE_ID}  # ID for load testing
  header_downstream -Server                       # Remove server header

  without /db
}
```
Here's the [documentation on Caddy's http.proxy](https://caddyserver.com/docs/proxy#syntax) directive, for further reference.  

It'd be great if we can knock out a high-performance example together - let me know what you think!

Reply:
So the issue isn't specifically the timeouts, it's that HAProxy currently closes out every request on the backend (i.e. HAProxy <-> CouchDB) when a request finishes, so there really isn't a reuseable connection pool at the moment.

I've personally seen some weird behaviour in production when people have used LB-managed connection pools under heavy load that cleared up when we disabled the connection pooling. I know for a fact it's especially bad if you use SSL between the reverse proxy (in the case I'm thinking of, it was an F5 BigIP) and CouchDB, but that has more to do with the poor implementation by Erlang of SSL. (We put a workaround into 2.2.0 for this via a `vm.args` config setting to auto-close any SSL connections after 300s.)

So the only change I'm recommending to the docs here is not to recommend backend connection pooling, when we recommend the exact opposite in `haproxy.cfg`. You could choose to leave out that part of the configuration, as I did with the Nginx config (the point was to provide a working example, not a tuned high-perf one), or we can hold up this PR until I have time to build a test harness and compare pooled vs. unpooled, which might be months.

My recommendation: remove the `timeout` line entirely. The Caddy default is 30s, which is fine, and if people want something higher, they can do that.

If you want to match the haproxy config more exactly, I would add `keepalive 0`.

Reply:
> I know for a fact it's especially bad if you use SSL between the reverse proxy... and CouchDB

Saw this as well, so I offload encryption to `stunnel` and trick the Couches into thinking all their cluster buddies are on loopback.

It looks like WireGuard will be pulled into the Linux kernel soon - how cool would it be if cluster setup bootstrapped its own VPN in the background...!

>My recommendation: remove the timeout line entirely. The Caddy default is 30s, which is fine, and if people want something higher, they can do that.
>
>If you want to match the haproxy config more exactly, I would add keepalive 0.

Implemented - https://github.com/apache/couchdb-documentation/pull/318/commits/74df1dc7a5dff3ecf35bd886b08daa054d1cc457  👍 Thanks for your detailed feedback, as always!